core/dive.c

```diff
@@ -489,14 +489,24 @@ static void copy_dc(const struct divecomputer *sdc, struct divecomputer *ddc)
 
 static void dc_cylinder_renumber(struct dive *dive, struct divecomputer *dc, const int mapping[]);
 
-/* copy dive computer and renumber the cylinders */
+/* copy dive computer list and renumber the cylinders
+ * space for the first divecomputer is provided by the
+ * caller, the remainder is allocated */
 static void copy_dc_renumber(struct dive *d, const struct divecomputer *sdc, struct divecomputer *ddc, const int cylinders_map[])
 {
-	*ddc = *sdc;
-	ddc->model = copy_string(sdc->model);
-	copy_samples(sdc, ddc);
-	copy_events(sdc, ddc);
-	dc_cylinder_renumber(d, ddc, cylinders_map);
+	for (;;) {
+		*ddc = *sdc;
+		ddc->model = copy_string(sdc->model);
+		copy_samples(sdc, ddc);
+		copy_events(sdc, ddc);
+		dc_cylinder_renumber(d, ddc, cylinders_map);
+		if (!sdc->next)
+			break;
+		sdc = sdc->next;
+		ddc->next = calloc(1, sizeof(struct divecomputer));
+		ddc = ddc->next;
+	}
+	ddc->next = NULL;
```

Comment:
This second version at least handles the "merge two and two" case right.

Reply:
Hmm. I see a different bug, but it's not new.

The

        *ddc = *sdc;

thing is wrong for the extradata. It happens to work (because I suspect we never change that list, and we never free it), but I think we should do "copy_extra_data()" the same way we do "copy_events()" etc.

Similarly, we do

        ddc->model = copy_string(sdc->model);

to copt the model string, but don't do the same for "serial" and "fw_version".

But as mentioned, this is all pre-existing, and the new copy_dc_renumber() function just duplicates the (mostly harmless) bug that existed in "copy_dc()".

So I think this commit looks fine.

Reply:
> but don't do the same for "serial" and "fw_version".

Yes, I noticed that. Then I realized that these fields are also not freed in free_dc(). So I assumed they were some global things. If this is not the case, this certainly should be cleaned up.

Reply:
I'm happy to entertain a follow up PR to clean up those more "harmless" issues.
Yes, we treat the strings in extra data like const (i.e., we never edit them and never free them), so because of that it's fine to share them, but it would be much more consistent to actually have hem private for each dive computer structure.
But I'd also love to see Travis finish on this one so we can get it merged.
I have some data downloading I want to do and that does merge into dives with multiple DCs...